File: xlogin.js
import fetch from "node-fetch";
import fs from "fs";
import { createInterface } from "readline";

const rl = createInterface({ input: process.stdin, output: process.stdout });
const prompt = (q) => new Promise(r => rl.question(q, a => r(a.trim())));

const BEARER = "AAAAAAAAAAAAAAAAAAAAANRILgAAAAAAnNwIzUejRCOuH5E6I8xnZz4puTs%3D1Zv7ttfk8LF81IUq16cHjhLTvJu4FA33AGWWjCpTnA";

async function getGuestToken() {
  const r = await fetch("https://api.twitter.com/1.1/guest/activate.json", {
    method: "POST",
    headers: {
      "authorization": `Bearer ${BEARER}`,
      "content-type": "application/json",
      "user-agent": "TwitterAndroid/10.21.0-release.0",
    },
  });
  const d = await r.json();
  return d.guest_token;
}

async function apiPost(path, body, guest, ct0, authToken, att) {
  const cookie = ct0 ? `ct0=${ct0}; auth_token=${authToken}` : "";
  const r = await fetch(`https://api.twitter.com${path}`, {
    method: "POST",
    headers: {
      "authorization": `Bearer ${BEARER}`,
      "content-type": "application/json",
      "user-agent": "TwitterAndroid/10.21.0-release.0",
      "x-twitter-active-user": "yes",
      "x-twitter-client-language": "en",
      "x-guest-token": guest || "",

      ...(ct0 ? { "x-csrf-token": ct0, "cookie": cookie } : {}),
    },
    body: JSON.stringify(body),
  });
  const text = await r.text();
  try {
    return { data: JSON.parse(text), headers: r.headers };
  } catch {
    throw new Error(`Response bukan JSON: ${text.slice(0, 100)}`);
  }
}

async function loginAccount(username, password) {
  console.log(`\n[${username}] Mulai login...`);

  const guest = await getGuestToken();
  console.log(`[${username}] ✅ Guest token: ${guest}`);

  // Flow init
  let { data } = await apiPost("/1.1/onboarding/task.json?flow_name=login", {
    input_flow_data: {
      flow_context: { debug_overrides: {}, start_location: { location: "splash_screen" } }
    },
    subtask_versions: {
      contacts_live_sync_permission_prompt: 0,
      email_verification: 2,
      topics_selector: 1,
      wait_spinner: 3,
      cta: 7
    }
  }, guest);

  let flowToken = data.flow_token;
  if (!flowToken) throw new Error(`No flow_token: ${JSON.stringify(data).slice(0,200)}`);

  // JS Instrumentation (required step)
  if (data.subtasks?.[0]?.subtask_id === "LoginJsInstrumentationSubtask") {
    ({ data } = await apiPost("/1.1/onboarding/task.json", {
      flow_token: flowToken,
      subtask_inputs: [{
        subtask_id: "LoginJsInstrumentationSubtask",
        js_instrumentation: { response: "{}", link: "next_link" }
      }]
    }, guest));
    flowToken = data.flow_token;
    console.log(`[${username}] JS Inst: ${JSON.stringify(data).slice(0,300)}`);
  }

  // Username
  ({ data } = await apiPost("/1.1/onboarding/task.json", {
    flow_token: flowToken,
    subtask_inputs: [{
      subtask_id: "LoginEnterUserIdentifierSSO",
      settings_list: {
        setting_responses: [{ key: "user_identifier", response_data: { text_data: { result: username } } }],
        link: "next_link"
      }
    }]
  }, guest));
  flowToken = data.flow_token;
  console.log(`[${username}] Username step: ${JSON.stringify(data).slice(0,300)}`);

  // Password
  ({ data } = await apiPost("/1.1/onboarding/task.json", {
    flow_token: flowToken,
    subtask_inputs: [{
      subtask_id: "LoginEnterPassword",
      enter_password: { password, link: "next_link" }
    }]
  }, guest));
  flowToken = data.flow_token;

  // Cek subtask
  let subtaskId = data.subtasks?.[0]?.subtask_id;
  console.log(`[${username}] Subtask: ${subtaskId}`);

  // OTP
  if (subtaskId === "LoginAcid") {
    const hint = data.subtasks[0]?.enter_text?.hint_text || "email/phone";
    console.log(`[${username}] ⚠️ Butuh verifikasi (${hint})`);
    const otp = await prompt(`[${username}] Masukkan OTP: `);
    ({ data } = await apiPost("/1.1/onboarding/task.json", {
      flow_token: flowToken,
      subtask_inputs: [{ subtask_id: "LoginAcid", enter_text: { text: otp, link: "next_link" } }]
    }, guest));
    flowToken = data.flow_token;
    subtaskId = data.subtasks?.[0]?.subtask_id;
  }

  // Duplication check
  if (subtaskId === "AccountDuplicationCheck") {
    ({ data } = await apiPost("/1.1/onboarding/task.json", {
      flow_token: flowToken,
      subtask_inputs: [{ subtask_id: "AccountDuplicationCheck", check_logged_in_account: { link: "AccountDuplicationCheck_false" } }]
    }, guest));
    flowToken = data.flow_token;
  }

  // Extract cookies dari response
  const setCookies = data.subtasks?.[0]?.open_account?.user?.id_str;
  
  // Coba ambil token dari cookies via endpoint verifikasi
  const verifyR = await fetch("https://api.twitter.com/1.1/account/verify_credentials.json", {
    headers: {
      "authorization": `Bearer ${BEARER}`,
      "x-guest-token": guest,
      "user-agent": "TwitterAndroid/10.21.0-release.0",
    }
  });

  // Parse cookies dari response headers
  const rawCookies = verifyR.headers.raw()["set-cookie"] || [];
  let auth_token = "", ct0 = "";
  for (const c of rawCookies) {
    if (c.includes("auth_token=")) auth_token = c.match(/auth_token=([^;]+)/)?.[1] || "";
    if (c.includes("ct0=")) ct0 = c.match(/ct0=([^;]+)/)?.[1] || "";
  }

  // Fallback: ambil dari open_account
  const openAccount = data.subtasks?.find(s => s.subtask_id === "LoginSuccessSubtask" || s.open_account);
  if (!auth_token && openAccount?.open_account) {
    auth_token = openAccount.open_account.oauth_token || "";
    ct0 = openAccount.open_account.oauth_token_secret || "";
  }

  if (!auth_token) {
    console.log(`[${username}] ❌ Gagal ambil auth_token`);
    console.log(`[${username}] Raw: ${JSON.stringify(data).slice(0, 300)}`);
    return null;
  }

  console.log(`[${username}] ✅ Login berhasil`);
  return { username, auth_token, ct0 };
}

async function main() {
  const raw = fs.readFileSync("twitter_accounts.txt", "utf-8").trim();
  const accounts = raw.split("\n\n").map(block => {
    const lines = block.trim().split("\n").map(l => l.trim());
    return { username: lines[0], password: lines[1] };
  });

  console.log(`Total akun: ${accounts.length}`);

  const results = [];
  for (const { username, password } of accounts) {
    try {
      const data = await loginAccount(username, password);
      if (data) results.push(data);
    } catch (e) {
      console.log(`[${username}] ❌ Error: ${e.message}`);
    }
  }

  fs.writeFileSync("twitter_tokens.json", JSON.stringify(results, null, 2));
  console.log(`\n✅ Selesai! ${results.length}/${accounts.length} berhasil`);
  console.log(`Tersimpan: twitter_tokens.json`);
  rl.close();
}

main().catch(console.error);

File: x_login.py
#!/usr/bin/env python3
"""
X (Twitter) Login Script
- Support single akun / semua akun dari accounts.txt
- Handle OTP/email verification
- Save session cookies ke folder sessions/
"""

import requests
import json
import os
import time

# ── Constants ──────────────────────────────────────────────────────────────────
BEARER_TOKEN = "AAAAAAAAAAAAAAAAAAAAANRILgAAAAAAnNwIzUejRCOuH5E6I8xnZz4puTs%3D1Zv7ttfk8LTMGsbGosIfl7vyFn8UCIUrsA17I8TzXDLCX2kbQ1XSB1dHQDFbA"

HEADERS_BASE = {
    "authorization": f"Bearer {BEARER_TOKEN}",
    "content-type": "application/json",
    "user-agent": "Mozilla/5.0 (Linux; Android 11; Termux) AppleWebKit/537.36",
    "x-twitter-active-user": "yes",
    "x-twitter-client-language": "en",
}

SESSIONS_DIR = "sessions"
ACCOUNTS_FILE = "accounts.txt"

# ── Session setup ──────────────────────────────────────────────────────────────
def make_session():
    s = requests.Session()
    s.headers.update(HEADERS_BASE)
    return s

def get_guest_token(s):
    r = s.post("https://api.twitter.com/1.1/guest/activate.json")
    r.raise_for_status()
    token = r.json()["guest_token"]
    s.headers["x-guest-token"] = token
    return token

# ── Login flow ─────────────────────────────────────────────────────────────────
def init_flow(s):
    payload = {
        "flow_name": "login",
        "input_flow_data": {
            "flow_context": {
                "debug_overrides": {},
                "start_location": {"location": "splash_screen"}
            }
        }
    }
    r = s.post(
        "https://api.twitter.com/1.1/onboarding/task.json?flow_name=login",
        json=payload
    )
    r.raise_for_status()
    return r.json()

def next_flow(s, flow_token, subtask_inputs):
    payload = {
        "flow_token": flow_token,
        "subtask_inputs": subtask_inputs
    }
    r = s.post("https://api.twitter.com/1.1/onboarding/task.json", json=payload)
    r.raise_for_status()
    return r.json()

def get_subtask_id(flow_resp):
    subtasks = flow_resp.get("subtasks", [])
    if subtasks:
        return subtasks[0].get("subtask_id", "")
    return ""

# ── Login steps ────────────────────────────────────────────────────────────────
def login(username, password):
    s = make_session()
    print(f"  [*] Getting guest token...")
    get_guest_token(s)

    print(f"  [*] Init login flow...")
    flow = init_flow(s)
    flow_token = flow["flow_token"]

    # Step 1: JS instrumentation
    print(f"  [*] Step 1: instrumentation...")
    flow = next_flow(s, flow_token, [{
        "subtask_id": "LoginJsInstrumentationSubtask",
        "js_instrumentation": {
            "response": json.dumps({"rf": {}, "s": ""}),
            "link": "next_link"
        }
    }])
    flow_token = flow["flow_token"]

    # Step 2: Enter username
    print(f"  [*] Step 2: username...")
    flow = next_flow(s, flow_token, [{
        "subtask_id": "LoginEnterUserIdentifierSSO",
        "settings_list": {
            "setting_responses": [{
                "key": "user_identifier",
                "response_data": {"text_data": {"result": username}}
            }],
            "link": "next_link"
        }
    }])
    flow_token = flow["flow_token"]
    subtask = get_subtask_id(flow)

    # Handle unusual activity check (kadang X minta username lagi)
    if subtask == "LoginEnterAlternateIdentifierSubtask":
        print(f"  [!] X minta verifikasi tambahan (username/email/phone):")
        alt = input("  Masukkan username/email/no HP terdaftar: ").strip()
        flow = next_flow(s, flow_token, [{
            "subtask_id": "LoginEnterAlternateIdentifierSubtask",
            "enter_text": {"text": alt, "link": "next_link"}
        }])
        flow_token = flow["flow_token"]

    # Step 3: Enter password
    print(f"  [*] Step 3: password...")
    flow = next_flow(s, flow_token, [{
        "subtask_id": "LoginEnterPassword",
        "enter_password": {"password": password, "link": "next_link"}
    }])
    flow_token = flow["flow_token"]
    subtask = get_subtask_id(flow)

    # Step 4: Handle OTP / email verification
    if subtask in ("LoginAcid", "ArkoseLogin"):
        print(f"  [!] OTP diminta!")
        otp = input("  Masukkan kode OTP dari email/SMS: ").strip()
        flow = next_flow(s, flow_token, [{
            "subtask_id": subtask,
            "enter_text": {"text": otp, "link": "next_link"}
        }])
        flow_token = flow["flow_token"]
        subtask = get_subtask_id(flow)

    # Step 5: Account duplication check
    if subtask == "AccountDuplicationCheck":
        flow = next_flow(s, flow_token, [{
            "subtask_id": "AccountDuplicationCheck",
            "check_logged_in_account": {"link": "AccountDuplicationCheck_false"}
        }])
        flow_token = flow["flow_token"]

    # Cek sukses
    subtask = get_subtask_id(flow)
    if subtask not in ("", None) and "LoginSuccess" not in subtask:
        print(f"  [?] Subtask tidak dikenal: {subtask}")
        print(f"  Response: {json.dumps(flow, indent=2)[:500]}")
        return None

    # Ambil auth token dari cookies
    cookies = dict(s.cookies)
    if "auth_token" not in cookies:
        print(f"  [!] Login gagal — auth_token tidak ditemukan")
        return None

    print(f"  [+] Login berhasil!")
    return cookies

# ── Save / load session ────────────────────────────────────────────────────────
def save_session(username, cookies):
    os.makedirs(SESSIONS_DIR, exist_ok=True)
    path = os.path.join(SESSIONS_DIR, f"{username}.json")
    with open(path, "w") as f:
        json.dump(cookies, f, indent=2)
    print(f"  [+] Session disimpan: {path}")

def load_session(username):
    path = os.path.join(SESSIONS_DIR, f"{username}.json")
    if os.path.exists(path):
        with open(path) as f:
            return json.load(f)
    return None

# ── Process accounts ───────────────────────────────────────────────────────────
def process_account(username, password):
    print(f"\n{'='*50}")
    print(f"Akun: @{username}")

    existing = load_session(username)
    if existing and "auth_token" in existing:
        print(f"  [~] Session sudah ada, skip login")
        return True

    try:
        cookies = login(username, password)
        if cookies:
            save_session(username, cookies)
            return True
        return False
    except requests.HTTPError as e:
        print(f"  [!] HTTP Error: {e.response.status_code} - {e.response.text[:200]}")
        return False
    except Exception as e:
        print(f"  [!] Error: {e}")
        return False

def load_accounts():
    if not os.path.exists(ACCOUNTS_FILE):
        print(f"[!] File {ACCOUNTS_FILE} tidak ditemukan!")
        print(f"    Buat file dengan format: username:password (satu per baris)")
        return []
    accounts = []
    with open(ACCOUNTS_FILE) as f:
        for line in f:
            line = line.strip()
            if line and ":" in line:
                parts = line.split(":", 1)
                accounts.append((parts[0].strip(), parts[1].strip()))
    return accounts

# ── Main ───────────────────────────────────────────────────────────────────────
def main():
    print("=== X Login Script ===")
    print(f"Sessions disimpan di: ./{SESSIONS_DIR}/\n")

    accounts = load_accounts()
    if not accounts:
        return

    print(f"Ditemukan {len(accounts)} akun di {ACCOUNTS_FILE}")
    print("\nPilih mode:")
    print("  1. Login 1 akun")
    print("  2. Login semua akun")
    choice = input("\nPilihan (1/2): ").strip()

    if choice == "1":
        print("\nDaftar akun:")
        for i, (u, _) in enumerate(accounts, 1):
            status = "✓ ada session" if load_session(u) else "  belum login"
            print(f"  {i}. @{u}  [{status}]")
        idx = input("\nPilih nomor akun: ").strip()
        try:
            idx = int(idx) - 1
            if 0 <= idx < len(accounts):
                u, p = accounts[idx]
                process_account(u, p)
            else:
                print("[!] Nomor tidak valid")
        except ValueError:
            print("[!] Input tidak valid")

    elif choice == "2":
        success = 0
        for u, p in accounts:
            ok = process_account(u, p)
            if ok:
                success += 1
            time.sleep(2)  # jeda antar akun biar gak rate limit
        print(f"\n{'='*50}")
        print(f"Selesai: {success}/{len(accounts)} akun berhasil")

    else:
        print("[!] Pilihan tidak valid")

if __name__ == "__main__":
    main()
